File: src/falcon_core/math/vector.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.vector import Vector as _CVector
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.communications.voltage_states.device_voltage_states import DeviceVoltageStates
from falcon_core.generic.list import List
from falcon_core.generic.list import List
from falcon_core.generic.list import List
from falcon_core.generic.map import Map
from falcon_core.generic.map import Map
from falcon_core.generic.pair import Pair
from falcon_core.math.point import Point
from falcon_core.physics.units.symbol_unit import SymbolUnit

class Vector:
    """Python wrapper for Vector."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, start: Point, end: Point) -> Vector:
        return cls(_CVector.new(start._c, end._c))

    @classmethod
    def new_from_end(cls, end: Point) -> Vector:
        return cls(_CVector.new_from_end(end._c))

    @classmethod
    def new_from_quantities(cls, start: Map, end: Map) -> Vector:
        return cls(_CVector.new_from_quantities(start._c, end._c))

    @classmethod
    def new_from_end_quantities(cls, end: Map) -> Vector:
        return cls(_CVector.new_from_end_quantities(end._c))

    @classmethod
    def new_from_doubles(cls, start: Map, end: Map, unit: SymbolUnit) -> Vector:
        return cls(_CVector.new_from_doubles(start._c, end._c, unit._c))

    @classmethod
    def new_from_end_doubles(cls, end: Map, unit: SymbolUnit) -> Vector:
        return cls(_CVector.new_from_end_doubles(end._c, unit._c))

    @classmethod
    def new_from_parent(cls, items: Map) -> Vector:
        return cls(_CVector.new_from_parent(items._c))

    @classmethod
    def from_json(cls, json: str) -> Vector:
        return cls(_CVector.from_json(json))

    def endPoint(self, ) -> Point:
        ret = self._c.endPoint()
        if ret is None: return None
        return Point._from_capi(ret)

    def startPoint(self, ) -> Point:
        ret = self._c.startPoint()
        if ret is None: return None
        return Point._from_capi(ret)

    def end_quantities(self, ) -> Map:
        ret = self._c.end_quantities()
        if ret is None: return None
        return Map(ret)

    def start_quantities(self, ) -> Map:
        ret = self._c.start_quantities()
        if ret is None: return None
        return Map(ret)

    def end_map(self, ) -> Map:
        ret = self._c.end_map()
        if ret is None: return None
        return Map(ret)

    def start_map(self, ) -> Map:
        ret = self._c.start_map()
        if ret is None: return None
        return Map(ret)

    def connections(self, ) -> List:
        ret = self._c.connections()
        if ret is None: return None
        return List(ret)

    def unit(self, ) -> SymbolUnit:
        ret = self._c.unit()
        if ret is None: return None
        return SymbolUnit._from_capi(ret)

    def principle_connection(self, ) -> Connection:
        ret = self._c.principle_connection()
        if ret is None: return None
        return Connection._from_capi(ret)

    def magnitude(self, ) -> None:
        ret = self._c.magnitude()
        return ret

    def insert_or_assign(self, key: Connection, value: Pair) -> None:
        ret = self._c.insert_or_assign(key._c, value._c)
        return ret

    def insert(self, key: Connection, value: Pair) -> None:
        ret = self._c.insert(key._c, value._c)
        return ret

    def at(self, key: Connection) -> Pair:
        ret = self._c.at(key._c)
        if ret is None: return None
        return Pair(ret)

    def erase(self, key: Connection) -> None:
        ret = self._c.erase(key._c)
        return ret

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def empty(self, ) -> None:
        ret = self._c.empty()
        return ret

    def clear(self, ) -> None:
        ret = self._c.clear()
        return ret

    def contains(self, key: Connection) -> None:
        ret = self._c.contains(key._c)
        return ret

    def keys(self, ) -> List:
        ret = self._c.keys()
        if ret is None: return None
        return List(ret)

    def values(self, ) -> List:
        ret = self._c.values()
        if ret is None: return None
        return List(ret)

    def items(self, ) -> List:
        ret = self._c.items()
        if ret is None: return None
        return List(ret)

    def addition(self, other: Vector) -> Vector:
        ret = self._c.addition(other._c)
        return cls._from_capi(ret)

    def subtraction(self, other: Vector) -> Vector:
        ret = self._c.subtraction(other._c)
        return cls._from_capi(ret)

    def double_multiplication(self, scalar: Any) -> Vector:
        ret = self._c.double_multiplication(scalar)
        return cls._from_capi(ret)

    def int_multiplication(self, scalar: Any) -> Vector:
        ret = self._c.int_multiplication(scalar)
        return cls._from_capi(ret)

    def double_division(self, scalar: Any) -> Vector:
        ret = self._c.double_division(scalar)
        return cls._from_capi(ret)

    def int_division(self, scalar: Any) -> Vector:
        ret = self._c.int_division(scalar)
        return cls._from_capi(ret)

    def negation(self, ) -> Vector:
        ret = self._c.negation()
        return cls._from_capi(ret)

    def update_start_from_states(self, state: DeviceVoltageStates) -> Vector:
        ret = self._c.update_start_from_states(state._c)
        return cls._from_capi(ret)

    def translate_doubles(self, point: Map, unit: SymbolUnit) -> Vector:
        ret = self._c.translate_doubles(point._c, unit._c)
        return cls._from_capi(ret)

    def translate_quantities(self, point: Map) -> Vector:
        ret = self._c.translate_quantities(point._c)
        return cls._from_capi(ret)

    def translate(self, point: Point) -> Vector:
        ret = self._c.translate(point._c)
        return cls._from_capi(ret)

    def translate_to_origin(self, ) -> Vector:
        ret = self._c.translate_to_origin()
        return cls._from_capi(ret)

    def double_extend(self, extension: Any) -> Vector:
        ret = self._c.double_extend(extension)
        return cls._from_capi(ret)

    def int_extend(self, extension: Any) -> Vector:
        ret = self._c.int_extend(extension)
        return cls._from_capi(ret)

    def double_shrink(self, extension: Any) -> Vector:
        ret = self._c.double_shrink(extension)
        return cls._from_capi(ret)

    def int_shrink(self, extension: Any) -> Vector:
        ret = self._c.int_shrink(extension)
        return cls._from_capi(ret)

    def unit_vector(self, ) -> Vector:
        ret = self._c.unit_vector()
        return cls._from_capi(ret)

    def normalize(self, ) -> Vector:
        ret = self._c.normalize()
        return cls._from_capi(ret)

    def project(self, other: Vector) -> Vector:
        ret = self._c.project(other._c)
        return cls._from_capi(ret)

    def update_unit(self, unit: SymbolUnit) -> None:
        ret = self._c.update_unit(unit._c)
        return ret

    def equal(self, b: Vector) -> None:
        ret = self._c.equal(b._c)
        return ret

    def not_equal(self, b: Vector) -> None:
        ret = self._c.not_equal(b._c)
        return ret

    def __add__(self, other):
        """Operator overload for +"""
        if isinstance(other, Vector):
            return self.addition(other)
        return NotImplemented

    def __sub__(self, other):
        """Operator overload for -"""
        if isinstance(other, Vector):
            return self.subtraction(other)
        return NotImplemented

    def __mul__(self, other):
        """Operator overload for *"""
        if isinstance(other, float):
            return self.double_multiplication(other)
        if isinstance(other, int):
            return self.int_multiplication(other)
        return NotImplemented

    def __truediv__(self, other):
        """Operator overload for /"""
        if isinstance(other, float):
            return self.double_division(other)
        if isinstance(other, int):
            return self.int_division(other)
        return NotImplemented

    def __neg__(self):
        """Operator overload for unary -"""
        return self.negation()

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Vector):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Vector):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/src/math/vector.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class c_vector (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.vector_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module Vector = struct
  type t = c_vector

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.vector_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.vector_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_vector ptr

  let make (start : Point.t) (end_ : Point.t) : t =
    ErrorHandling.multi_read [start; end_] (fun () ->
      let ptr = Capi_bindings.vector_create start#raw end_#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromEnd (end_ : Point.t) : t =
    ErrorHandling.read end_ (fun () ->
      let ptr = Capi_bindings.vector_create_from_end end_#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromQuantities (start : Mapconnectionquantity.t) (end_ : Mapconnectionquantity.t) : t =
    ErrorHandling.multi_read [start; end_] (fun () ->
      let ptr = Capi_bindings.vector_create_from_quantities start#raw end_#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromEndQuantities (end_ : Mapconnectionquantity.t) : t =
    ErrorHandling.read end_ (fun () ->
      let ptr = Capi_bindings.vector_create_from_end_quantities end_#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromDoubles (start : Mapconnectiondouble.t) (end_ : Mapconnectiondouble.t) (unit : Symbolunit.t) : t =
    ErrorHandling.multi_read [start; end_; unit] (fun () ->
      let ptr = Capi_bindings.vector_create_from_doubles start#raw end_#raw unit#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromEndDoubles (end_ : Mapconnectiondouble.t) (unit : Symbolunit.t) : t =
    ErrorHandling.multi_read [end_; unit] (fun () ->
      let ptr = Capi_bindings.vector_create_from_end_doubles end_#raw unit#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let fromParent (items : Mapconnectionquantity.t) : t =
    ErrorHandling.read items (fun () ->
      let ptr = Capi_bindings.vector_create_from_parent items#raw in
      ErrorHandling.raise_if_error ();
      new c_vector ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.vector_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.vector_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let endPoint (handle : t) : Point.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_end_point handle#raw in
      ErrorHandling.raise_if_error ();
      new c_point result
    )

  let startPoint (handle : t) : Point.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_start_point handle#raw in
      ErrorHandling.raise_if_error ();
      new c_point result
    )

  let endQuantities (handle : t) : Mapconnectionquantity.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_end_quantities handle#raw in
      ErrorHandling.raise_if_error ();
      new c_mapconnectionquantity result
    )

  let startQuantities (handle : t) : Mapconnectionquantity.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_start_quantities handle#raw in
      ErrorHandling.raise_if_error ();
      new c_mapconnectionquantity result
    )

  let endMap (handle : t) : Mapconnectiondouble.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_end_map handle#raw in
      ErrorHandling.raise_if_error ();
      new c_mapconnectiondouble result
    )

  let startMap (handle : t) : Mapconnectiondouble.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_start_map handle#raw in
      ErrorHandling.raise_if_error ();
      new c_mapconnectiondouble result
    )

  let connections (handle : t) : Listconnection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_connections handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listconnection result
    )

  let unit (handle : t) : Symbolunit.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_unit handle#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let principleConnection (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_principle_connection handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let magnitude (handle : t) : float =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_magnitude handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let insertOrAssign (handle : t) (key : Connection.t) (value : Pairquantityquantity.t) : unit =
    ErrorHandling.multi_read [handle; key; value] (fun () ->
      let result = Capi_bindings.vector_insert_or_assign handle#raw key#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let insert (handle : t) (key : Connection.t) (value : Pairquantityquantity.t) : unit =
    ErrorHandling.multi_read [handle; key; value] (fun () ->
      let result = Capi_bindings.vector_insert handle#raw key#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let at (handle : t) (key : Connection.t) : Pairquantityquantity.t =
    ErrorHandling.multi_read [handle; key] (fun () ->
      let result = Capi_bindings.vector_at handle#raw key#raw in
      ErrorHandling.raise_if_error ();
      new c_pairquantityquantity result
    )

  let erase (handle : t) (key : Connection.t) : unit =
    ErrorHandling.multi_read [handle; key] (fun () ->
      let result = Capi_bindings.vector_erase handle#raw key#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let size (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_size handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let empty (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_empty handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let clear (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_clear handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let contains (handle : t) (key : Connection.t) : bool =
    ErrorHandling.multi_read [handle; key] (fun () ->
      let result = Capi_bindings.vector_contains handle#raw key#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let keys (handle : t) : Listconnection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_keys handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listconnection result
    )

  let values (handle : t) : Listpairquantityquantity.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_values handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listpairquantityquantity result
    )

  let items (handle : t) : Listpairconnectionpairquantityquantity.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_items handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listpairconnectionpairquantityquantity result
    )

  let addition (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.vector_addition handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let subtraction (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.vector_subtraction handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let doubleMultiplication (handle : t) (scalar : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_double_multiplication handle#raw scalar in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let intMultiplication (handle : t) (scalar : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_int_multiplication handle#raw scalar in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let doubleDivision (handle : t) (scalar : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_double_division handle#raw scalar in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let intDivision (handle : t) (scalar : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_int_division handle#raw scalar in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let negation (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_negation handle#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let updateStartFromStates (handle : t) (state : Devicevoltagestates.t) : t =
    ErrorHandling.multi_read [handle; state] (fun () ->
      let result = Capi_bindings.vector_update_start_from_states handle#raw state#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let translateDoubles (handle : t) (point : Mapconnectiondouble.t) (unit : Symbolunit.t) : t =
    ErrorHandling.multi_read [handle; point; unit] (fun () ->
      let result = Capi_bindings.vector_translate_doubles handle#raw point#raw unit#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let translateQuantities (handle : t) (point : Mapconnectionquantity.t) : t =
    ErrorHandling.multi_read [handle; point] (fun () ->
      let result = Capi_bindings.vector_translate_quantities handle#raw point#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let translate (handle : t) (point : Point.t) : t =
    ErrorHandling.multi_read [handle; point] (fun () ->
      let result = Capi_bindings.vector_translate handle#raw point#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let translateToOrigin (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_translate_to_origin handle#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let doubleExtend (handle : t) (extension : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_double_extend handle#raw extension in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let intExtend (handle : t) (extension : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_int_extend handle#raw extension in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let doubleShrink (handle : t) (extension : float) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_double_shrink handle#raw extension in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let intShrink (handle : t) (extension : int) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_int_shrink handle#raw extension in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let unitVector (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_unit_vector handle#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let normalize (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.vector_normalize handle#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let project (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.vector_project handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_vector result
    )

  let updateUnit (handle : t) (unit : Symbolunit.t) : unit =
    ErrorHandling.multi_read [handle; unit] (fun () ->
      let result = Capi_bindings.vector_update_unit handle#raw unit#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end